github/prci.py

```diff
@@ -19,13 +20,27 @@
 )
 from internals.gql import util, queries
 
-
 logger = logging.getLogger(__name__)
 
 
 ERROR_BACKOFF_TIME = 600
 
 
+def initial_cleanup():
+    """
+        The main idea of this function is to clean stucked vagrant process
+        and remove the vagrant vms before a new run.
+    """
+    stuck_process = "ps aux | grep -i '[v]agrant up' | awk '{ print $2 }' | xargs kill -9"
+    cmd = "vagrant global-status --prune | cut -f1 -d ' ' | sed -e '/^$/,$d'| sed '1d;2d' | xargs vagrant destroy --force"
```

Comment:
What happens when this code runs is the following:

- systemd calls from prci.service unit and the ExecStart line is calling for bash, that is calling python via shebang defined in `prci.py` file;
- Python is spawning two new processes via `subprocess`, one for each execution.
- I'm not configuring `subprocess` to capture output (arg `capture_output=True`), that means the output is not captured inside python `subprocess` but is passed to outside bash caller, created by systemd.

```
If capture_output is true, stdout and stderr will be captured. When used, the internal Popen object is automatically created with stdout=PIPE and stderr=PIPE. The stdout and stderr arguments may not be supplied at the same time as capture_output.
```

- Per default, any system unit started by systemd will redirect STDOUT and STDERR to journald:

``` bash
-- Reboot --
jun 18 14:53:41 reproducer-3130.lab.cerberus.domain systemd[1]: Started FreeIPA PR CI.
jun 18 14:53:42 reproducer-3130.lab.cerberus.domain bash[1163]: kill: not enough arguments
jun 18 14:53:45 reproducer-3130.lab.cerberus.domain bash[1163]: A Vagrant environment or target machine is required to run this
jun 18 14:53:45 reproducer-3130.lab.cerberus.domain bash[1163]: command. Run `vagrant init` to create a new Vagrant environment. Or,
jun 18 14:53:45 reproducer-3130.lab.cerberus.domain bash[1163]: get an ID of a target machine from `vagrant global-status` to run
jun 18 14:53:45 reproducer-3130.lab.cerberus.domain bash[1163]: this command on. A final option is to change to a directory with a
jun 18 14:53:45 reproducer-3130.lab.cerberus.domain bash[1163]: Vagrantfile and to try again.
jun 18 14:53:45 reproducer-3130.lab.cerberus.domain bash[1163]: /root/freeipa-pr-ci/github/prci.py:59: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml>
jun 18 14:53:45 reproducer-3130.lab.cerberus.domain bash[1163]:   return yaml.load(yml_file)
jun 18 14:53:46 reproducer-3130.lab.cerberus.domain bash[1163]: /root/freeipa-pr-ci/github/internals/entities.py:500: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read htt>
jun 18 14:53:46 reproducer-3130.lab.cerberus.domain bash[1163]:   return yaml.load(tasks_file_content)["jobs"]
jun 18 14:53:46 reproducer-3130.lab.cerberus.domain bash[1163]: INFO: __main__: Task fedora-29/build on PR #6 is stale. Updating for rerun.
jun 18 14:53:54 reproducer-3130.lab.cerberus.domain bash[1163]: INFO: __main__: Skipping fedora-29/simple_replication of #6: waiting for dependencies
jun 18 14:53:54 reproducer-3130.lab.cerberus.domain bash[1163]: INFO: __main__: Skipping fedora-29/external_ca_1 of #6: not enough resources
jun 18 14:53:54 reproducer-3130.lab.cerberus.domain bash[1163]: INFO: __main__: Skipping fedora-29/external_ca_2 of #6: waiting for dependencies
jun 18 14:53:54 reproducer-3130.lab.cerberus.domain bash[1163]: INFO: __main__: Skipping fedora-29/test_topologies of #6: waiting for dependencies
jun 18 14:53:54 reproducer-3130.lab.cerberus.domain bash[1163]: INFO: __main__: Skipping fedora-29/test_sudo of #6: not enough resources
jun 18 14:53:54 reproducer-3130.lab.cerberus.domain bash[1163]: INFO: __main__: Skipping fedora-29/test_commands of #6: waiting for dependencies
```

- If we run this manually, this is the result:

``` bash
[root@reproducer-3130 ~]# vagrant global-status
id       name    provider state     directory
--------------------------------------------------------------------------------------------------
57a6719  builder libvirt preparing /root/freeipa-pr-ci/jobs/2eb8626a-91f2-11e9-99c4-001c42026ead

The above shows information about all known Vagrant environments
on this machine. This data is cached and may not be completely
up-to-date (use "vagrant global-status --prune" to prune invalid
entries). To interact with any of the machines, you can go to that
directory and run Vagrant, or you can use the ID directly with
Vagrant commands from any directory. For example:
"vagrant destroy 1a2b3c4d"
[root@reproducer-3130 ~]# systemctl status prci
● prci.service - FreeIPA PR CI
   Loaded: loaded (/etc/systemd/system/prci.service; enabled; vendor preset: disabled)
   Active: active (running) since Tue 2019-06-18 14:53:41 -03; 14min ago
 Main PID: 1163 (prci.py)
    Tasks: 7 (limit: 4915)
   Memory: 152.0M
   CGroup: /system.slice/prci.service
           ├─1163 /usr/bin/python3 /root/freeipa-pr-ci/github/prci.py reproducer-3130 --config /root/.config/freeipa-pr-ci/config.yml
           └─1329 /usr/bin/ruby-mri /usr/share/vagrant/gems/bin/vagrant up --no-provision --parallel

jun 18 14:54:58 reproducer-3130.lab.cerberus.domain bash[1163]: DEBUG: root: ==> builder:  -- Graphics Password: Not defined
jun 18 14:54:58 reproducer-3130.lab.cerberus.domain bash[1163]: DEBUG: root: ==> builder:  -- Video Type:        cirrus
jun 18 14:54:58 reproducer-3130.lab.cerberus.domain bash[1163]: DEBUG: root: ==> builder:  -- Video VRAM:        9216
jun 18 14:54:58 reproducer-3130.lab.cerberus.domain bash[1163]: DEBUG: root: ==> builder:  -- Sound Type:
jun 18 14:54:58 reproducer-3130.lab.cerberus.domain bash[1163]: DEBUG: root: ==> builder:  -- Keymap:            en-us
jun 18 14:54:58 reproducer-3130.lab.cerberus.domain bash[1163]: DEBUG: root: ==> builder:  -- TPM Path:
jun 18 14:54:58 reproducer-3130.lab.cerberus.domain bash[1163]: DEBUG: root: ==> builder:  -- INPUT:             type=mouse, bus=ps2
jun 18 14:54:59 reproducer-3130.lab.cerberus.domain bash[1163]: DEBUG: root: ==> builder: Creating shared folders metadata...
jun 18 14:54:59 reproducer-3130.lab.cerberus.domain bash[1163]: DEBUG: root: ==> builder: Starting domain.
jun 18 14:54:59 reproducer-3130.lab.cerberus.domain bash[1163]: DEBUG: root: ==> builder: Waiting for domain to get an IP address...
[root@reproducer-3130 ~]# systemctl stop prci
^C
[root@reproducer-3130 ~]# ps aux | grep -i prci
root      1163  0.0  0.5 485860 47324 ?        Ssl  14:53   0:00 /usr/bin/python3 /root/freeipa-pr-ci/github/prci.py reproducer-3130 --config /root/.config/freeipa-pr-ci/config.yml
root      1909  0.0  0.0 215744   828 pts/1    S+   15:08   0:00 grep --color=auto -i prci
[root@reproducer-3130 ~]# kill -9 1163
[root@reproducer-3130 ~]# systemctl status prci
● prci.service - FreeIPA PR CI
   Loaded: loaded (/etc/systemd/system/prci.service; enabled; vendor preset: disabled)
   Active: failed (Result: signal) since Tue 2019-06-18 15:09:01 -03; 5s ago
  Process: 1163 ExecStart=/bin/bash -c PYTHONPATH=$PYTHONPATH:/root/freeipa-pr-ci /root/freeipa-pr-ci/github/prci.py "$(hostname -s)" --config /root/.config/freeipa-pr-ci/config.yml (code=killed, signal=KILL)
 Main PID: 1163 (code=killed, signal=KILL)

jun 18 15:08:13 reproducer-3130.lab.cerberus.domain bash[1163]: DEBUG: root: ==> builder:  -- Sound Type:
jun 18 15:08:13 reproducer-3130.lab.cerberus.domain bash[1163]: DEBUG: root: ==> builder:  -- Keymap:            en-us
jun 18 15:08:13 reproducer-3130.lab.cerberus.domain bash[1163]: DEBUG: root: ==> builder:  -- TPM Path:
jun 18 15:08:13 reproducer-3130.lab.cerberus.domain bash[1163]: DEBUG: root: ==> builder:  -- INPUT:             type=mouse, bus=ps2
jun 18 15:08:13 reproducer-3130.lab.cerberus.domain bash[1163]: DEBUG: root: ==> builder: Creating shared folders metadata...
jun 18 15:08:13 reproducer-3130.lab.cerberus.domain bash[1163]: DEBUG: root: ==> builder: Starting domain.
jun 18 15:08:14 reproducer-3130.lab.cerberus.domain bash[1163]: DEBUG: root: ==> builder: Waiting for domain to get an IP address...
jun 18 15:09:00 reproducer-3130.lab.cerberus.domain systemd[1]: prci.service: Main process exited, code=killed, status=9/KILL
jun 18 15:09:01 reproducer-3130.lab.cerberus.domain systemd[1]: prci.service: Failed with result 'signal'.
jun 18 15:09:01 reproducer-3130.lab.cerberus.domain systemd[1]: Stopped FreeIPA PR CI.
[root@reproducer-3130 ~]# vagrant global-status

id       name    provider state     directory
--------------------------------------------------------------------------------------------------
0941350  builder libvirt preparing /root/freeipa-pr-ci/jobs/2eb8626a-91f2-11e9-99c4-001c42026ead

The above shows information about all known Vagrant environments
on this machine. This data is cached and may not be completely
up-to-date (use "vagrant global-status --prune" to prune invalid
entries). To interact with any of the machines, you can go to that
directory and run Vagrant, or you can use the ID directly with
Vagrant commands from any directory. For example:
"vagrant destroy 1a2b3c4d"
[root@reproducer-3130 ~]#
[root@reproducer-3130 ~]# vagrant global-status --prune | cut -f1 -d ' ' | sed -e '/^$/,$d'| sed '1d;2d' | xargs vagrant destroy --force
==> builder: Removing domain...
[root@reproducer-3130 ~]# echo $?
0
[root@reproducer-3130 ~]#  vagrant global-status --prune | cut -f1 -d ' ' | sed -e '/^$/,$d'| sed '1d;2d' | xargs vagrant destroy --force
A Vagrant environment or target machine is required to run this
command. Run `vagrant init` to create a new Vagrant environment. Or,
get an ID of a target machine from `vagrant global-status` to run
this command on. A final option is to change to a directory with a
Vagrantfile and to try again.
[root@reproducer-3130 ~]# echo $?
123
[root@reproducer-3130 ~]#
```

Reply:
More content about process invocation:

- In early steps of start, we can see the processes doing the cleanup:

```
[root@reproducer-3130 ~]# systemctl status prci
● prci.service - FreeIPA PR CI
   Loaded: loaded (/etc/systemd/system/prci.service; enabled; vendor preset: disabled)
   Active: active (running) since Tue 2019-06-18 16:12:12 -03; 2s ago
 Main PID: 2393 (prci.py)
    Tasks: 8 (limit: 4915)
   Memory: 90.3M
   CGroup: /system.slice/prci.service
           ├─2393 /usr/bin/python3 /root/freeipa-pr-ci/github/prci.py reproducer-3130 --config /root/.config/freeipa-pr-ci/config.yml
           ├─2403 /bin/sh -c vagrant global-status --prune | cut -f1 -d ' ' | sed -e '/^$/,$d'| sed '1d;2d' | xargs vagrant destroy --force
           ├─2408 xargs vagrant destroy --force
           └─2431 /usr/bin/ruby-mri /usr/share/vagrant/gems/bin/vagrant destroy --force f68c366
```